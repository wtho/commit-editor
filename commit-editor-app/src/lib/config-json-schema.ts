import type { Monaco } from './monaco'

type ValueType =
  | { type: 'array'; items: { type: 'string' } }
  | { type: 'string' | 'integer' }
  | { enum: readonly (string | number)[] }
  | {}

const ruleType = ({
  description,
  title,
  ...valueType
}: ValueType & { description?: string; title?: string }) => {
  const defaultRuleTuple = [{ enum: [0, 1, 2] }, { enum: ['always', 'never'] }]
  const hasKeys = Object.keys(valueType).length > 0
  return {
    type: 'array',
    items: hasKeys ? [...defaultRuleTuple, valueType] : defaultRuleTuple,
    description,
    title,
  }
}

export const caseArray = [
  'lower-case',
  'upper-case',
  'camel-case',
  'kebab-case',
  'pascal-case',
  'sentence-case',
  'snake-case',
  'start-case',
] as const
export type Case = typeof caseArray[number]
const caseDescription = `
'lower-case': default
'upper-case': UPPERCASE
'camel-case': camelCase
'kebab-case': kebab-case
'pascal-case': PascalCase
'sentence-case': Sentence case
'snake-case': snake_case
'start-case': Start Case`

export function initConfigSchema(monaco: Monaco) {
  monaco.languages.json.jsonDefaults.setDiagnosticsOptions({
    validate: true,
    schemaValidation: 'error',
    schemas: [
      {
        uri: 'internal://server/config.json',
        fileMatch: ['config.json'],
        schema: {
          $schema: 'http://json-schema.org/draft-07/schema#',
          $id: 'internal://server/config.json',
          type: 'object',
          properties: {
            extends: {
              type: 'array',
              items: {
                enum: [
                  '@commitlint/config-conventional',
                  '@commitlint/config-angular',
                  '@commitlint/config-angular-type-enum',
                ],
              },
            },
            // TODO: parserPreset
            // TODO: formatter
            // TODO: ignores
            defaultIgnores: {
              type: 'boolean',
            },
            rules: {
              properties: {
                'body-full-stop': ruleType({
                  type: 'string',
                  title: 'body ends with value',
                  description: `The body must end with the configured value, e. g. '.'`,
                }),
                'body-leading-blank': ruleType({
                  title: 'body begins with blank line',
                  description: `The line before the body must be a blank line`,
                }),
                'body-empty': ruleType({
                  title: 'body is empty',
                  description: `The body must kept empty`,
                }),
                'body-max-length': ruleType({
                  type: 'integer',
                  title: 'body has value or less characters',
                  description:
                    'The body can consist of only up to the number of characters as configured',
                }), // TODO check infinity in json
                'body-max-line-length': ruleType({
                  type: 'integer',
                  title: 'body lines has value or less characters',
                  description: `Each line in the body can have a maximum length of characters as configured`,
                }),
                'body-min-length': ruleType({
                  type: 'integer',
                  title: 'body has value or more characters',
                  description: `The body must consist of at least as many characters as configured`,
                }),
                'body-case': ruleType({
                  enum: caseArray,
                  title: 'body is in case value',
                  description: `The body must be written in the configured case:${caseDescription}`,
                }),
                'footer-leading-blank': ruleType({
                  title: 'footer begins with blank line',
                  description: `The line before the footer must be a blank line`,
                }),
                'footer-empty': ruleType({
                  title: 'footer is empty',
                  description: `A footer is not allowed, it must be empty`,
                }),
                'footer-max-length': ruleType({
                  type: 'integer',
                  title: 'footer has value or less characters',
                  description: `The footer can consist of only up to the number of characters as configured`,
                }),
                'footer-max-line-length': ruleType({
                  type: 'integer',
                  title: 'footer lines has value or less characters',
                  description: `Each line in the footer can have a maximum length of characters as configured`,
                }),
                'footer-min-length': ruleType({
                  type: 'integer',
                  title: 'footer has value or more characters',
                  description: `The footer must consist of at least as many characters as configured`,
                }),
                'header-case': ruleType({
                  enum: caseArray,
                  title: 'header is in case value',
                  description: `The header must be written in the configured case:${caseDescription}`,
                }),
                'header-full-stop': ruleType({
                  type: 'string',
                  title: 'header ends with value',
                  description: `The header must end with the configured value, e. g. '.'`,
                }),
                'header-max-length': ruleType({
                  type: 'integer',
                  title: 'header has value or less characters',
                  description: `The header can consist of only up to the number of characters as configured`,
                }),
                'header-min-length': ruleType({
                  type: 'integer',
                  title: 'header has value or more characters',
                  description: `The header must consist of at least as many characters as configured`,
                }),
                'references-empty': ruleType({
                  title: 'references has at least one entry',
                  description: `When configured with 'never', the references must have at least one entry`,
                }),
                'scope-enum': ruleType({
                  type: 'array',
                  items: { type: 'string' },
                  title: 'scope is found in value',
                  description: `The scope value in brackets after the type must be one of the specified values`,
                }),
                'scope-case': ruleType({
                  enum: caseArray,
                  title: 'scope is in case value',
                  description: `The scope must be written in the configured case:${caseDescription}`,
                }),
                'scope-empty': ruleType({
                  title: 'scope is empty',
                  description: `A scope value is not allowed, it must be empty`,
                }),
                'scope-max-length': ruleType({
                  type: 'integer',
                  title: 'scope has value or less characters',
                  description: `The scope can consist of only up to the number of characters as configured`,
                }),
                'scope-min-length': ruleType({
                  type: 'integer',
                  title: 'scope has value or more characters',
                  description: `The scope must consist of at least as many characters as configured`,
                }),
                'subject-case': ruleType({
                  enum: caseArray,
                  title: 'subject is in case value',
                  description: `The subject must be written in the configured case:${caseDescription}`,
                }),
                'subject-empty': ruleType({
                  title: 'subject is empty',
                  description: `A subject is not allowed, it must be empty`,
                }),
                'subject-full-stop': ruleType({
                  type: 'string',
                  title: 'subject ends with value',
                  description: `The subject must end with the configured value, e. g. '.'`,
                }),
                'subject-max-length': ruleType({
                  type: 'integer',
                  title: 'subject has value or less characters',
                  description: `The subject can consist of only up to the number of characters as configured`,
                }),
                'subject-min-length': ruleType({
                  type: 'integer',
                  title: 'subject has value or more characters',
                  description: `The subject must consist of at least as many characters as configured`,
                }),
                'type-enum': ruleType({
                  type: 'array',
                  items: { type: 'string' },
                  title: 'type is found in value',
                  description: `The type value must be one of the specified values`,
                }),
                'type-case': ruleType({
                  enum: caseArray,
                  title: 'type is in case value',
                  description: `The type must be written in the configured case:${caseDescription}`,
                }),
                'type-empty': ruleType({
                  title: 'type is empty',
                  description: `Type is not allowed, it must be empty`,
                }),
                'type-max-length': ruleType({
                  type: 'integer',
                  title: 'type has value or less characters',
                  description: `The type can consist of only up to the number of characters as configured`,
                }),
                'type-min-length': ruleType({
                  type: 'integer',
                  title: 'type has value or more characters',
                  description: `The type must consist of at least as many characters as configured`,
                }),
                'signed-off-by': ruleType({
                  type: 'string',
                  title: 'message has value',
                  description: `message must contain the configured value, e. g. 'Signed-off-by:'`,
                }),
              },
            },
          },
          additionalProperties: false,
        },
      },
    ],
  })
}
