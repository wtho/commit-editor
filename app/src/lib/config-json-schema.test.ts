import { initConfigSchema } from './config-json-schema'

describe('config-json-schema.ts', () => {
  test('should init config schema', () => {

    const setDiagnosticsOptions = jest.fn()
    const monacoMock = {
      languages: {
        json: {
          jsonDefaults: {
            setDiagnosticsOptions
          }
        }
      }
    }

    initConfigSchema(monacoMock as any)

    expect(setDiagnosticsOptions).toHaveBeenCalled()
  })

  test('should snap config schema', () => {
    const setDiagnosticsOptions = jest.fn()
    const monacoMock = {
      languages: {
        json: {
          jsonDefaults: {
            setDiagnosticsOptions
          }
        }
      }
    }

    initConfigSchema(monacoMock as any)

    const schema = setDiagnosticsOptions.mock.calls[0][0].schemas[0].schema
    expect(schema).toMatchSnapshot()
  })
})