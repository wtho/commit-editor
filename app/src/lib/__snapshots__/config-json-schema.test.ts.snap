// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config-json-schema.ts should snap config schema 1`] = `
Object {
  "$id": "internal://server/config.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": Object {
    "defaultIgnores": Object {
      "type": "boolean",
    },
    "extends": Object {
      "items": Object {
        "enum": Array [
          "@commitlint/config-conventional",
          "@commitlint/config-angular",
          "@commitlint/config-angular-type-enum",
        ],
      },
      "type": "array",
    },
    "rules": Object {
      "properties": Object {
        "body-case": Object {
          "description": "The body must be written in the configured case:
'lower-case': default
'upper-case': UPPERCASE
'camel-case': camelCase
'kebab-case': kebab-case
'pascal-case': PascalCase
'sentence-case': Sentence case
'snake-case': snake_case
'start-case': Start Case",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "enum": Array [
                "lower-case",
                "upper-case",
                "camel-case",
                "kebab-case",
                "pascal-case",
                "sentence-case",
                "snake-case",
                "start-case",
              ],
            },
          ],
          "title": "body is in case value",
          "type": "array",
        },
        "body-empty": Object {
          "description": "The body must kept empty",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
          ],
          "title": "body is empty",
          "type": "array",
        },
        "body-full-stop": Object {
          "description": "The body must end with the configured value, e. g. '.'",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "string",
            },
          ],
          "title": "body ends with value",
          "type": "array",
        },
        "body-leading-blank": Object {
          "description": "The line before the body must be a blank line",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
          ],
          "title": "body begins with blank line",
          "type": "array",
        },
        "body-max-length": Object {
          "description": "The body can consist of only up to the number of characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "body has value or less characters",
          "type": "array",
        },
        "body-max-line-length": Object {
          "description": "Each line in the body can have a maximum length of characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "body lines has value or less characters",
          "type": "array",
        },
        "body-min-length": Object {
          "description": "The body must consist of at least as many characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "body has value or more characters",
          "type": "array",
        },
        "footer-empty": Object {
          "description": "A footer is not allowed, it must be empty",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
          ],
          "title": "footer is empty",
          "type": "array",
        },
        "footer-leading-blank": Object {
          "description": "The line before the footer must be a blank line",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
          ],
          "title": "footer begins with blank line",
          "type": "array",
        },
        "footer-max-length": Object {
          "description": "The footer can consist of only up to the number of characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "footer has value or less characters",
          "type": "array",
        },
        "footer-max-line-length": Object {
          "description": "Each line in the footer can have a maximum length of characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "footer lines has value or less characters",
          "type": "array",
        },
        "footer-min-length": Object {
          "description": "The footer must consist of at least as many characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "footer has value or more characters",
          "type": "array",
        },
        "header-case": Object {
          "description": "The header must be written in the configured case:
'lower-case': default
'upper-case': UPPERCASE
'camel-case': camelCase
'kebab-case': kebab-case
'pascal-case': PascalCase
'sentence-case': Sentence case
'snake-case': snake_case
'start-case': Start Case",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "enum": Array [
                "lower-case",
                "upper-case",
                "camel-case",
                "kebab-case",
                "pascal-case",
                "sentence-case",
                "snake-case",
                "start-case",
              ],
            },
          ],
          "title": "header is in case value",
          "type": "array",
        },
        "header-full-stop": Object {
          "description": "The header must end with the configured value, e. g. '.'",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "string",
            },
          ],
          "title": "header ends with value",
          "type": "array",
        },
        "header-max-length": Object {
          "description": "The header can consist of only up to the number of characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "header has value or less characters",
          "type": "array",
        },
        "header-min-length": Object {
          "description": "The header must consist of at least as many characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "header has value or more characters",
          "type": "array",
        },
        "references-empty": Object {
          "description": "When configured with 'never', the references must have at least one entry",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
          ],
          "title": "references has at least one entry",
          "type": "array",
        },
        "scope-case": Object {
          "description": "The scope must be written in the configured case:
'lower-case': default
'upper-case': UPPERCASE
'camel-case': camelCase
'kebab-case': kebab-case
'pascal-case': PascalCase
'sentence-case': Sentence case
'snake-case': snake_case
'start-case': Start Case",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "enum": Array [
                "lower-case",
                "upper-case",
                "camel-case",
                "kebab-case",
                "pascal-case",
                "sentence-case",
                "snake-case",
                "start-case",
              ],
            },
          ],
          "title": "scope is in case value",
          "type": "array",
        },
        "scope-empty": Object {
          "description": "A scope value is not allowed, it must be empty",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
          ],
          "title": "scope is empty",
          "type": "array",
        },
        "scope-enum": Object {
          "description": "The scope value in brackets after the type must be one of the specified values",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "items": Object {
                "type": "string",
              },
              "type": "array",
            },
          ],
          "title": "scope is found in value",
          "type": "array",
        },
        "scope-max-length": Object {
          "description": "The scope can consist of only up to the number of characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "scope has value or less characters",
          "type": "array",
        },
        "scope-min-length": Object {
          "description": "The scope must consist of at least as many characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "scope has value or more characters",
          "type": "array",
        },
        "signed-off-by": Object {
          "description": "message must contain the configured value, e. g. 'Signed-off-by:'",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "string",
            },
          ],
          "title": "message has value",
          "type": "array",
        },
        "subject-case": Object {
          "description": "The subject must be written in the configured case:
'lower-case': default
'upper-case': UPPERCASE
'camel-case': camelCase
'kebab-case': kebab-case
'pascal-case': PascalCase
'sentence-case': Sentence case
'snake-case': snake_case
'start-case': Start Case",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "enum": Array [
                "lower-case",
                "upper-case",
                "camel-case",
                "kebab-case",
                "pascal-case",
                "sentence-case",
                "snake-case",
                "start-case",
              ],
            },
          ],
          "title": "subject is in case value",
          "type": "array",
        },
        "subject-empty": Object {
          "description": "A subject is not allowed, it must be empty",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
          ],
          "title": "subject is empty",
          "type": "array",
        },
        "subject-full-stop": Object {
          "description": "The subject must end with the configured value, e. g. '.'",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "string",
            },
          ],
          "title": "subject ends with value",
          "type": "array",
        },
        "subject-max-length": Object {
          "description": "The subject can consist of only up to the number of characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "subject has value or less characters",
          "type": "array",
        },
        "subject-min-length": Object {
          "description": "The subject must consist of at least as many characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "subject has value or more characters",
          "type": "array",
        },
        "type-case": Object {
          "description": "The type must be written in the configured case:
'lower-case': default
'upper-case': UPPERCASE
'camel-case': camelCase
'kebab-case': kebab-case
'pascal-case': PascalCase
'sentence-case': Sentence case
'snake-case': snake_case
'start-case': Start Case",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "enum": Array [
                "lower-case",
                "upper-case",
                "camel-case",
                "kebab-case",
                "pascal-case",
                "sentence-case",
                "snake-case",
                "start-case",
              ],
            },
          ],
          "title": "type is in case value",
          "type": "array",
        },
        "type-empty": Object {
          "description": "Type is not allowed, it must be empty",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
          ],
          "title": "type is empty",
          "type": "array",
        },
        "type-enum": Object {
          "description": "The type value must be one of the specified values",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "items": Object {
                "type": "string",
              },
              "type": "array",
            },
          ],
          "title": "type is found in value",
          "type": "array",
        },
        "type-max-length": Object {
          "description": "The type can consist of only up to the number of characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "type has value or less characters",
          "type": "array",
        },
        "type-min-length": Object {
          "description": "The type must consist of at least as many characters as configured",
          "items": Array [
            Object {
              "enum": Array [
                0,
                1,
                2,
              ],
            },
            Object {
              "enum": Array [
                "always",
                "never",
              ],
            },
            Object {
              "type": "integer",
            },
          ],
          "title": "type has value or more characters",
          "type": "array",
        },
      },
    },
  },
  "type": "object",
}
`;
